src/serialization/deserializer.rs

```diff
@@ -87,31 +86,30 @@ impl Deserializer {
         let mut payload_slice = &mut self.payload_buf[0..payload_len];
         reader.read_exact(&mut payload_slice)?;
 
-        let mut cursor = Cursor::new(&payload_slice);
-        let mut dest_index: usize = 0;
+        let mut payload_index: usize = 0;
         let mut restat_state = RestatState::new();
-        while cursor.position() < payload_slice.len() as u64 {
-            let num = zig_zag_decode(varint_read(&mut cursor)?);
-
-            if num < 0 {
-                let zero_count = (-num).to_usize()
-                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
-                // skip the zeros
-                dest_index = dest_index.checked_add(zero_count)
-                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
-                continue;
-            } else {
-                let count: T = T::from_i64(num)
-                    .ok_or(DeserializeError::UnsuitableCounterType)?;
-
-                h.set_count_at_index(dest_index, count)
-                    .map_err(|_| DeserializeError::EncodedArrayTooLong)?;
-
-                restat_state.on_nonzero_count(dest_index, count);
-
-                dest_index = dest_index.checked_add(1)
-                    .ok_or(DeserializeError::UsizeTypeTooSmall)?;
-            }
+        let mut decode_state = DecodeLoopState::new();
+
+        while payload_index < payload_len.saturating_sub(9) {
```

Comment:
I assume the compiler is smart enough to hoist the saturating sub out of the loop.